src/control/system/faultdomain_test.go

```diff
@@ -0,0 +1,510 @@
+		},
+		"whitespace-only strings": {
+			input:  []string{"ok", "\t    "},
+			expErr: errors.New("invalid fault domain"),
+		},
+		"single-level": {
+			input: []string{"ok"},
+			expResult: &FaultDomain{
+				domains: []string{"ok"},
+			},
+		},
+		"multi-level": {
+			input: []string{"ok", "go"},
+			expResult: &FaultDomain{
+				domains: []string{"ok", "go"},
+			},
+		},
+		"trim whitespace": {
+			input: []string{" ok  ", "\tgo\n"},
+			expResult: &FaultDomain{
+				domains: []string{"ok", "go"},
+			},
+		},
+		"fix case": {
+			input: []string{"OK", "Go"},
+			expResult: &FaultDomain{
+				domains: []string{"ok", "go"},
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			result, err := NewFaultDomain(tc.input...)
+
+			common.CmpErr(t, tc.expErr, err)
+
+			if diff := cmp.Diff(result, tc.expResult, cmp.AllowUnexported(FaultDomain{})); diff != "" {
+				t.Fatalf("(-want, +got): %s", diff)
+			}
+		})
+	}
+}
+
+func TestSystem_NewFaultDomainFromString(t *testing.T) {
+	for name, tc := range map[string]struct {
+		input     string
+		expErr    error
+		expResult *FaultDomain
+	}{
+		"empty": {
+			expResult: &FaultDomain{},
+		},
+		"only whitespace": {
+			input:     "  \n  \t",
+			expResult: &FaultDomain{},
+		},
+		"single-level fault domain": {
+			input: "/host",
+			expResult: &FaultDomain{
+				domains: []string{"host"},
+			},
+		},
+		"multi-level fault domain": {
+			input: "/dc0/rack1/pdu2/host",
+			expResult: &FaultDomain{
+				domains: []string{"dc0", "rack1", "pdu2", "host"},
+			},
+		},
+		"fix whitespace errors": {
+			input: " / rack 1/pdu 0    /host\n",
+			expResult: &FaultDomain{
+				domains: []string{"rack 1", "pdu 0", "host"},
+			},
+		},
+		"symbols ok": {
+			input: "/$$$/#/!?/--++/}{",
+			expResult: &FaultDomain{
+				domains: []string{"$$$", "#", "!?", "--++", "}{"},
+			},
+		},
+		"case insensitive": {
+			input: "/DC0/Rack1/PDU2/Host",
+			expResult: &FaultDomain{
+				domains: []string{"dc0", "rack1", "pdu2", "host"},
+			},
+		},
+		"fault domain doesn't start with separator": {
+			input:  "junk",
+			expErr: errors.New("invalid fault domain"),
+		},
+		"fault domain ends with separator": {
+			input:  "/junk/",
+			expErr: errors.New("invalid fault domain"),
+		},
+		"only root separator": {
+			input:  "/",
+			expErr: errors.New("invalid fault domain"),
+		},
+		"fault domain with empty levels": {
+			input:  "/dc0//pdu2/host",
+			expErr: errors.New("invalid fault domain"),
+		},
+		"fault domain with only whitespace between levels": {
+			input:  "/dc0/    /pdu2/host",
+			expErr: errors.New("invalid fault domain"),
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			result, err := NewFaultDomainFromString(tc.input)
+
+			common.CmpErr(t, tc.expErr, err)
+
+			if diff := cmp.Diff(result, tc.expResult, cmp.AllowUnexported(FaultDomain{})); diff != "" {
+				t.Fatalf("(-want, +got): %s", diff)
+			}
+		})
+	}
+}
+
+func TestSystem_FaultDomain_String(t *testing.T) {
```

Comment:
Not necessarily anything wrong with doing it this way, but you could combine the previous suite with this one and just compare the output of .String() to the input string to make sure that they match. Just a bit less maintenance overhead.

Reply:
Hm, but in some cases we normalize the input string. I guess we could add an expStr to the test case struct in the previous test. I think keeping them separate may be a bit easier to understand, though.

Reply:
Yeah that's fine. The test is already written and I wouldn't expect these tests to need a lot of maintenance.